### ISSUE TITLE
Set $modx, $modx->user, and $modx->resource early in Setup

### ISSUE THREAD

**Issue by BobRay** [modxcms/revolution#9030]
bobray created Redmine issue ID 9030

If an upgrade saves a user or resource and a plugin connected to an invoked event calls get() on one of these, Setup will crash if these are not set.

This affects an early 2.2.x upgrade.

If possible, xPDO::OPT_SETUP should also be set earlier in Setup, because it's set too late to help with this problem.

**Comment by Mark-H** [comment]
Always check if you received what you expect from any API. `if ($modx->getUser() && $modx->resource)`.

**Comment by BobRay** [comment]
The point is that a plugin with a missing sanity check shouldn't cause setup to crash. The user has no idea what's happened, little hope of fixing things,  and lost confidence in MODX. 

I don't think there's any reason for plugins to be executing during setup. 

If xPDO::OPT_SETUP were set early, the code that runs plugins could be altered to skip them if xPDO::OPT_SETUP is set.

**Comment by OptimusCrime** [comment]
Can we just turn them off without any side effects?

**Comment by Mark-H** [comment]
It's still seems like a developer error to me, but sure I'll reopen if you'd like to send a pull request to fix it.

**Comment by BobRay** [comment]
It *is * a developer error, but not one the user should suffer for, imo.

The fix is simple and I've done a PR. There will be a tiny performance hit, but it's a constant comparison, so it should be negligible.

**Comment by pixelchutes** [comment]
I recently experienced an error during setup for the first time from a plugin, and ended up tracing it down to pdoTools. I had to disable the plugin and clear all of core/cache/* in order to proceed.

+1 for the recommended approach in the PR and for preventing unnecessary events during setup.

**Comment by OptimusCrime** [comment]
Feels like a hack, and I've never encountered this before. However, if it solves a real problem and I'm +1 for implementing this. Great work with sending the pR @BobRay :)

**Comment by pbowyer** [comment]
+1. Can we merge and close? #modxbughunt

**Comment by Mark-H** [comment]
The pull request is not complete yet. When it's merged the issue will be closed too.

**Comment by BobRay** [comment]
I edited the PR. I hope it's correct.

**Comment by opengeek** [comment]
Are we sure this is the proper way to proceed? It seems to me there are some pretty valid scenarios where you would want plugins to execute during setup, especially if they are attached to resources and users.

**Comment by BobRay** [comment]
Excellent point. 

I filed this because a plugin of mine crashed setup. IIRC, it was a call to a method of $modx->user or $modx->resource (e.g., get() ) when the object didn't exist yet. I fixed my plugin with a simple sanity check, and it's a very rare problem, so maybe it's better to just fix the plugins when this occurs. 

@pixelchutes do you know what happened in the pdoTools plugin that crashed things?

**Comment by pixelchutes** [comment]
@bobray Unfortunately, I never got to the bottom of it, but I may have logged the exact error (I will have to check.) If I had to guess, I think the cache clear was the more likely fix vs. disabling the plugin. I say this because I have yet to experience the error since, and have kept it enabled.

NOTE: I did change from a development domain to a production domain when it happened (advanced upgrade on top of itself.) My belief is pdoTools parser had cached a path that no longer​ existed.

Thinking through this some more, both seem like edge cases/user errors, and I think we could consider closing this.

Thoughts?
